src/renderer/tr_bsp.cpp

```diff
@@ -230,16 +248,53 @@ static	void R_LoadLightmaps( lump_t *l, const char *psMapName ) {
 				image[j*4+3] = 255;
 			}
 		}
-		tr.lightmaps[i] = R_CreateImage( va("*%s/lightmap%d",sMapName,i), image,
-										 LIGHTMAP_SIZE, LIGHTMAP_SIZE, qfalse, qfalse,
-										 (qboolean) !!r_ext_compressed_lightmaps->integer, GL_CLAMP );
+
+		// vanilla hardcoded the maximum number of lightmaps to 256
+		// ideally, we use a 2048x2048 atlas to store all of those in one texture
+		int currentAtlas = i / LIGHTMAPS_PER_ATLAS;
+		int currentAtlasLightmap = i % LIGHTMAPS_PER_ATLAS;
+		int xoff = (currentAtlasLightmap % LIGHTMAPS_PER_ATLAS_DIMENSION) * LIGHTMAP_SIZE;
+		int yoff = (currentAtlasLightmap / LIGHTMAPS_PER_ATLAS_DIMENSION) * LIGHTMAP_SIZE;
+
+		// bind atlas for current lightmap and update the atlas with the lightmap data
+		GL_Bind(tr.lightmaps[currentAtlas]);
+		qglTexSubImage2D(
```

Comment:
Went another route, which I hope is ok. It will just fallback to vertex lit if glConfig.maxTextureSize is smaller than the atlas size and warn the user about that.